### DISCUSSION TITLE
autobegin, or "BEGIN (implicit)" preventing the use of "async with session.begin():"

### DISCUSSION THREAD

**Discussion by synic** [sqlalchemy/sqlalchemy#6864] [Usage Questions] [answered]
I am using SQLAlchemy 1.4.22.

I am having a bit of trouble figuring out how sessions are supposed to work when it comes to transactions. I have a fastapi application, where a connection is made, and a session is returned using a dependency, something like this:

```python
engine = create_async_engine(database_url)
SessionFactory = sessionmaker(bind=engine, class_=AsyncSession)
async_session = async_scoped_session(SessionFactory, scopefunc=asyncio.current_task)

async def get_sesion() -> AsyncSession:
    async with async_session() as session:
        yield session
```

The first time `get_session()` is used is in authentication. No `session.begin()` is used because no writing happens, and no transaction is needed. Once authentication is done, I have a route for adding a "brand" that looks like this:

```python
@app.route("/brands/", response_model=models.Brand)
async def add_brand(session: AsyncSession = Depends(get_session), payload: schemas.Brand):
    async with session.begin():
        brand = crud.add_brand(session, payload)

    return brand
```

The "async with session.begin()" fails with the following error:

```
  File "/app/./lr/brands/routes/brands.py", line 44, in add
    async with session.begin():
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/base.py", line 57, in __aenter__
    return await self.start(is_ctxmanager=True)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/session.py", line 441, in start
    await greenlet_spawn(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 116, in greenlet_spawn
    result = context.switch(*args, **kwargs)
  File "<string>", line 2, in begin
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 298, in warned
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1325, in begin
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.
```

I believe this is because during auth, a transaction is automatically created. When echoing, I can see this:

```
2021-08-06 16:38:05,400 INFO sqlalchemy.engine.Engine select version()
2021-08-06 16:38:05,400 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-08-06 16:38:05,401 INFO sqlalchemy.engine.Engine select current_schema()
2021-08-06 16:38:05,401 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-08-06 16:38:05,402 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2021-08-06 16:38:05,402 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-08-06 16:38:05,403 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-08-06 16:38:05,438 INFO sqlalchemy.engine.Engine SELECT users.id FROM users WHERE users.email = %s
2021-08-06 16:38:05,438 INFO sqlalchemy.engine.Engine [generated in 0.00020s] ('test@test.com',)
```

What is the correct way to do this? Do I need to be managing a transaction manually for the read-only part of the request? Is there a way to disable the new "autobegin" feature?

**Comment by CaselIT** [comment, marked as answer]
Please look at this https://github.com/sqlalchemy/sqlalchemy/discussions/6844

**Reply by synic** [reply]
Thank you!

I put it into our own subclass. 


```python
from contextlib import asynccontextmanager
from typing import AsyncGenerator

from sqlalchemy.ext.asyncio import AsyncSession as AsyncSessionBase


class AsyncSession(AsyncSessionBase):
    @asynccontextmanager
    async def commit_context(self) -> AsyncGenerator[None, None]:
        """Context manager to for committing a transaction.

        At the begging of the scope of the context, a transaction will be started,
        unless one is already in progress (which is most likely the case).

        When the scope of the context ends, a commit will be attempted. If any error
        happens within the scope of the context, the transaction will be rolled back and
        an error will be raised.

        This is, in essence, the same as ``async with session.begin():``, except that it
        will not throw an error if a transaction is already in process.

        See
        https://github.com/sqlalchemy/sqlalchemy/discussions/6864#discussioncomment-1142958
        for more information.
        """
        if not self.in_transaction():
            async with self.begin():
                yield
        else:
            yield
            if self.in_transaction():
                try:
                    await self.commit()
                except Exception:
                    await self.rollback()
                    raise
```

Avoided calling it `transaction` because I believe that is misleading. It is not a grouping of operations in a transaction, rather, it is a grouping of a scope that when it ends, the current transaction will be committed, and if any error happens in during those operations (including the commit), the transaction will be rolled back and an exception raised.

**Comment by zzzeek** [comment]
"autobegin" is not actually new, the `Session` has always begun a "transaction" immediately when it was first created.  what's different is that 1. it waits until it actually needs to before it starts this "transaction" and 2. you can now actually call `Session.begin()` to start this transaction *sooner* than this newly delayed start would normally happen.

the reason "transaction" is in quotes is because the Session's "transaction" is not linked to any database connection when first created.    it's basically a placeholder for the work that the Session will be doing, it fires off some events that lots of frameworks like zope.transaction rely upon, and that's it.     then as the Session is asked to do work on specific database connections, these connections are procured from the connection pool and then added to the state of this transaction object; it's usually just one connection but if the session is bound to multiple engines then it can be multiple connections.

before 1.4, you just didnt call session.begin() period unless you were using the "autocommit" mode of use, which didn't help anyone since it led to worse patterns.

the main decision you need to make here is regarding the checkout of a database connection.  if you want your authentication and "adding a brand" use cases to just use different transactions / checkouts, the simplest thing would be to call `session.close()` at the end of your authentication method, that will close out the connection in use and evict any objects associated with the session.    then the session will be in a brand new state for your "adding a brand" method.

alternatively, what the vast majority of applications do here is that they simply stick to basic [commit as you go](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#opening-and-closing-a-session) use, they let the "begin()" happen automatically like it always has and they just commit() when they've just made some data mutations they want persisted.  this would have your "auth" and "brand" methods work on the same DB connection within the same transaction and is usually fine.

on top of all of this is the important notion that SQLAlchemy builds upon the [pep-249](https://docs.sqlalchemy.org/en/14/glossary.html?highlight=pep%20249#term-DBAPI) model which importantly **does not have any begin method**.  Transactions in the DBAPI are always started **implicitly** unless the connection itself is put into a non-standard [autocommit mode](https://docs.sqlalchemy.org/en/14/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit) ahead of time.     So SQLAlchemy's Core actually has no code that emits "BEGIN" on database connections of any kind, outside of the realm of DBAPI emulation layers.

You're probably using asyncpg which means you actually **are** using such a DBAPI emulation layer, as asyncpg is not by itself a pep-249 library.  SQLAlchemy provides an emulation layer that allows it to implicitly begin, as the entirety of the rest of SQLAlchemy is built on this assumption.  Even the other async drivers we currently support (aiomysql, aiosqlite) have the "implicit begin" thing happening.

**Reply by synic** [reply]
Thank you.

**Comment by mmuraro** [comment]
Hi, I'm joining the discussion, maybe you can help me. My code should sequentially do the following operations on a PostgreSQL database:
- deleting data from the table (TRUNCATE)
- select query to verify the outcome of the modification (SELECT COUNT(*))
- inserting data into the table
- if all operations have been performed successfully COMMIT otherwise ROLLBACK
Currently the verify select query remains "IDLE IN TRANSACTION" and blocks the next insert.
If I COMMIT or ROLLBACK this also affects the TRUNCATE operation and I don't want this.
Suggestions?
A thousand thanks

**Reply by zzzeek** [reply]
> If I COMMIT or ROLLBACK this also affects the TRUNCATE operation and I don't want this.

What do you want then?   why TRUNCATE and not COMMIT ?

> Currently the verify select query remains "IDLE IN TRANSACTION" and blocks the next insert.

you need to end the transaction to continue.  this is not a SQLAlchemy thing, this is just how databases work

**Reply by mmuraro** [reply]
I would like to perform COMMIT or ROLLBACK only after carrying out the INSERT operations and not after the SELECT

**Reply by CaselIT** [reply]
how do you plan to achieve this?

> If I COMMIT or ROLLBACK this also affects the TRUNCATE operation and I don't want this.

it's not possible with a single transaction to only commit / rollback part of it, unless you want to use savepoints
